Add URI fragment as a separate combinator

swamp-agr opened the pull request:
## Fragment 

I would like to introduce `Fragment` combinator as URI Fragment.

```haskell
data Fragment' (mods :: [*]) (a :: *)
    deriving Typeable

type Fragment = Fragment' '[Lenient]
```

(updated. see review for more details)

It represents the URI fragment that quite useful in HTTP, e.g. as HTML anchor, despite it is not a part of REST API spec.

<img width="601" alt="87881364-6ebed680-ca01-11ea-927a-1b699e963760" src="https://user-images.githubusercontent.com/6595049/88487252-12712f00-cf8c-11ea-8a2a-8a1e2f31547e.png">

My implementation also provides following properties:
- Fragment should only have one type parameter (that should fit `ToHttpApiData`/`FromHttpApiData` constraints).
- Fragment should be present only once per API endpoint (implemented via type class and type families).
- Fragment should not impact API (i.e. server should return 200, `Nothing` should substitute in case of parsing failed).

## Motivation

Consider GitHub, e.g. following URL: https://github.com/haskell-servant/servant/blob/master/servant/src/Servant/API/QueryParam.hs#L46-L50

While the content of the page with and without anchor remains the same, the page markup is different. 

Imagine the situation where I want to introduce:
- server-side rendering for HTML page with similar (highlighting) functionality based on fragment.
- client-side rendering (e.g. custom browser capabilities) with the same motivation.

## Usage

Only one fragment per endpoint will be accepted.
```haskell
type PostTrackingApi = "post" :> QueryParam TrackingId :> Fragment Text :> Get '[HTML] Tracking
```

However, if user will specify several fragments per endpoint it will be rejected:
```haskell
>>> type FailExampleAPI = Fragment Bool :> ((Fragment Int :> Get '[HTML] NoContent) :<|> Post '[HTML] NoContent)
>>> instance OnlyOneFragment FailExampleAPI
<interactive>:109:10: error:
    • Only one Fragment allowed per endpoint in api ‘Fragment Bool
                                                     :> ((Fragment Int :> Get '[HTML] NoContent)
                                                         :<|> Post '[HTML] NoContent)’.
```


Closes #1322.

## Implementation details

1. [X] Fragment data type
2. [X] HasLink
   1. [X] MkLink
   2. [X] toLink
3. [X] Link
4. [X] Link -> URI
5. [X] [3/3] servant-server: HasServer
   1. [X] ServerT
   2. [X] route
   3. [X] hoistServerWithContext
6. [X] [2/2] servant-client-core
   1. [X] [3/3] HasClient
      1. [X] Client
      2. [X] clientHasRoute
      3. [X] hoistClientMonad
   2. [X] [6/6] RequestF
      1. [X] data type
7. [X] servant-client: defaultMakeClientRequest
8. [X] [4/4] after figuring type families: only one fragment per endpoint!
   1. [X] FragmentNotIn type family
   2. [X] FragmentUnique type family
   3. [X] TypeError on result of (5)
   4. [X] OnlyOneFragment type class
9. [X] [4/4] tests
    1. [X] doctest
    2. [X] [8/8] servant-docs: 
       1. [X] HasDocs
       2. [X] ToFragment
       3. [X] DocFragment
       4. [X] Lens' fragment
       5. [X] Action
       6. [X] defAction
       7. [X] combineAction
       8. [X] golden
    3. [X] [6/6] servant-foreign: HasForeign
       1. [X] extend Url with Maybe Frag
       2. [X] Frag data type
       3. [X] defUrl
       4. [X] lenses
       5. [X] prisms
       6. [X] HasForeign
    4. [X] servant-server:test:ServerSpec
        - [X] Links spec (URI fixed).
        - [X] Client spec.
        - [X] Server spec ("fragment absence should not impact API" part of URI fragment covered as well).

Please point me out if I am missing something.

swamp-agr commented:
Documentation added in db384303e2ce7b81e57ceec359b66dfb38771d12.

swamp-agr commented:
- Tests added in eda41d91f8153088a4a3ecd60519d1d24bb918a5:
    - [X] Links spec (URI fixed).
    - [X] Client spec.
    - [X] Server spec ("fragment absence should not impact API" part of URI fragment covered as well).

swamp-agr commented:
@fisx please provide your feedback on PR!

fisx commented:
> @fisx please provide your feedback on PR!

I can see you, and I like the idea, but I'll need a little more time.  #1314 goes first, then I'm a bit fuzzy about whose in line next.

I haven't checked, but this seems like something you could provide in a separate package for now?  That would probably be the fastest way into hackage, but I'm definitely interested.

Anybody else with the power of maintainership: feel free to steal this!  :)

swamp-agr commented:
I did not know about "pipeline" existence. 

- This feature is 100% ready as far as I can see.
- For sure, I will wait as long as needed and will try to keep it up-to-date (in corresponding order).
- Feel free to use check-list from here for #1314. I think, it is partially applicable and might be helpful.

Please let me know, once you or any other contributors are ready to let this feature in.

Thank you!

fisx commented:
There isn't really a pipeline, except for in my head.  Thanks for the code, it's really appreciated, I'll do my best to merge it as soon as I can!

fizruk commented on servant-docs/src/Servant/Docs/Internal.hs:
```diff
@@ -161,6 +161,22 @@ data DocQueryParam = DocQueryParam
+  , _fragDesc   :: String -- user supplied
+  } deriving (Eq, Ord, Show)
+
+-- | Combine two Fragments, we can't make a monoid because merging Fragments breaks
+-- the laws.
+--
+-- As such, we invent a non-commutative, left associative operation
+-- 'combineFragment' to mush two together taking the info from the very left.
+combineFragment :: Maybe DocFragment -> Maybe DocFragment -> Maybe DocFragment
+mdocFragment `combineFragment` _ = mdocFragment
```
1. Picking first occurrence is a valid monoid (see `First` in `Data.Monoid`).
2. I think ``Nothing `combineFragment` Just doc`` should be `Just doc`: if there was no fragment on the left, we should pick the fragment from the right.

```suggestion
-- | There should be at most one `Fragment` per API endpoint.
-- So here we are keeping the first occurrence.
combineFragment :: Maybe DocFragment -> Maybe DocFragment -> Maybe DocFragment
Nothing `combineFragment` mdocFragment = mdocFragment
Just docFragment `combineFragment` _ = Just docFragment
```

swamp-agr replied:
Fixed.

fizruk commented on servant-server/src/Servant/Server/Internal.hs:
```diff
@@ -852,5 +856,53 @@ type HasServerArrowTypeError a b =
+    let parseFragment = join . fmap (either (const Nothing) Just)
+          . join . fmap (either (const Nothing) Just)
+          . fmap (fmap parseQueryParam . T.decodeUtf8')
+          . identifyFragment (not . BC8.null) (crop . rawPathInfo) (crop . rawQueryString)
```
I think instead of `(not . BC8.null)` we should check the presence of `#`. The difference is that `/my/endpoint#` will have an empty fragment while `/my/endpoint` will have no fragment.

swamp-agr replied:
Fixed.

fizruk commented on servant/src/Servant/API/Fragment.hs:
```diff
@@ -0,0 +1,26 @@
+--
+-- >>> -- /post#TRACKING
+-- >>> type MyApi = "post" :> Fragment Text :> Get '[JSON] Tracking
+data Fragment (a :: *)
```
`Fragment` behaves similarly to `QueryParam`, so I think we should add modifiers. I think that we should at least recognise `Lenient`/`Strict` and `Description` as we do for `QueryParam`. I am not sure if `Required` makes sense here, but we could also add `Required`/`Optional`.

swamp-agr replied:
- Modifiers added.
- Server restricted to `FoldLenient`.
- Fragment itself set by default to `Lenient` to allow user control of parse errors handling.

swamp-agr commented on servant-server/src/Servant/Server/Internal.hs:
```diff
@@ -852,5 +856,53 @@ type HasServerArrowTypeError a b =
+    let parseFragment = join . fmap (either (const Nothing) Just)
+          . join . fmap (either (const Nothing) Just)
+          . fmap (fmap parseQueryParam . T.decodeUtf8')
+          . identifyFragment (not . BC8.null) (crop . rawPathInfo) (crop . rawQueryString)
```
Fixed.

swamp-agr commented:
All comments from Review iteration 1 were resolved three days ago. 

- `Fragment' mods a`  introduced. 
- `Fragment` becomes `Lenient` by default.
- `"#"` and `""` now distinguished.

swamp-agr commented:
Feature is completed and still waiting for review.
All changes requested by @fizruk implemented month ago. Please provide your review.

@fisx Could we go ahead of UVerb and proceed with this feature?

arianvp commented:
I was always under the assumption that browsers strip the fragment before sending requests to the server. So I don't see why we would have to handle this in servant.
 Is that assumption wrong?

Your motivating example (Github lines) seems to indeed confirm my suspicion. If I click on the link; the fragment is never sent to the server:
![image](https://user-images.githubusercontent.com/628387/94807041-4daa3700-03ef-11eb-9c65-f898db8aa1ff.png)

Instead. Github parses the fragment in javascript and does the highlighting there.

arianvp commented:
I guess it's useful for the `HasLink` instances; you you can generate type-safe links to fragments

arianvp commented:
Indeed HTTP standard makes not mention of fragments in the URI syntax See https://ietf.org/rfc/rfc2616.html#section-3.2.2  and they should be stripped before sending to the server.

I'm ok with having the `Fragment` type merged as a way to generate type-safe links and docs with fragments in them.   But we'd need to make a few changes to this PR to make this spec-complaint.

1. `servant-server` should simply ignore the fragment; and not attempt to parse it. As HTTP-1.1 compliant servers don't know about fragment syntax.  

2. `servant-client` should not include the fragment in the request, as HTT-1.1 compliant clients never include the fragment in the Request URI but strip it on the client-side.

arianvp commented:
I'm happy to be convinced that is too much of a pedantic take by the way.   Overall the code looks good, and I don't see why we wouldn't merge this.  

@swamp-agr I also agree we can merge this before the Uverb things. They don't seem to influence eachother too much. And I'm happy to rebase the Uverb branch if needed. Lets not get that blocking you.  

Let me know what you think about the "Dont send fragments over the wire" part.  And after that lets get this merged ASAP

swamp-agr commented:
Hi @arianvp 

> Instead. Github parses the fragment in javascript and does the highlighting there.
> 
> From https://en.wikipedia.org/wiki/URI_fragment:
> > The fragment identifier functions differently to the rest of the URI: its processing is exclusively client-sided with no participation from the web server

Consider,
- Server-side rendering capabilities for HTML 
- and limited JavaScript enabled on the page as well. 
- Under these circumstances, fragment still might be helpful BUT should be reached and processed by server. 
- And this particular case seems to break traditional meaning of URI Fragment,
- i.e. code-highlighting could be recognised on server via fragment.
- Meantime, I realise that the case is extremely specific.

That is why I add `HasClient` and `HasServer` into the scope of changes in order to handle the fragment and allow user to decide what to do with it even when parsing failed on server side.
I realise that this behaviour is not aligned with the standard.

Please let me know what do you think about this specific case? 
If it looks like something really wrong, I will go ahead and restrict `HasClient` and `HasServer` changes as you requested.

Best Regards,
@swamp-agr

maksbotan commented:
Hello,

I would argue that adding server-side parsing of fragments and client-side sending of them is not the right thing to do, as that contradicts the HTTP spec, and common usage, and behavior of most clients. See also this discussion: https://github.com/OAI/OpenAPI-Specification/issues/1635

Particularly, note that `curl` strips fragment:

```
$ curl -v https://github.com/haskell-servant/servant/blob/master/servant/src/Servant/API/QueryParam.hs#L46-L50
...
> GET /haskell-servant/servant/blob/master/servant/src/Servant/API/QueryParam.hs HTTP/1.1
> Host: github.com
> User-Agent: curl/7.64.1
> Accept: */*
...
```

On the other hand, `ToLink` stuff may be useful to let users denote their usage for fragments for whatever reason.

In other words, I agree with @arianvp: let's leave only `ToLink` part of this PR. What do you think?

swamp-agr commented:
Hi @arianvp, @maksbotan 
- I absolutely agree that standard should not be violated. 
- You provided enough arguments to make me work around use-case above in another way. 
- I implemented following changes per your requests:

- [X] rebase over `0.18.1`.
- [X] servant-server:
   1. [X] src:
      1. [X] include HTTP spec reference into haddock.
      2. [X] replace current implementation with `skip`. i.e. server should ignore fragment.
      3. [X] restore `Delayed`.
   2. [X] test: servant should ignore `Fragment`.
- [X] servant-client-core:
   1. [X] src:
      1. [X] include HTTP spec reference into haddock.
      2. [X] replace current implementation with `skip`, i.e. client should ignore fragment.
      3. [X] restore `RequestF`.
- [X] servant-client:
   1. [X] src: restore `defaultMakeRequest`.
   2. [X] test: servant should ignore `Fragment`.

maksbotan commented on servant-client-core/src/Servant/Client/Core/Request.hs:
```diff
@@ -52,9 +52,10 @@ import           Network.HTTP.Types
                  http11, methodGet)
 import           Servant.API
-                 (ToHttpApiData, toEncodedUrlPiece, toHeader, SourceIO)
+                 (SourceIO, ToHttpApiData, toEncodedUrlPiece, toHeader)
```
Hm, is changing this file really neccessary? Generally, I'd only do formatting in files where I introduce significant changes.

swamp-agr replied:
`stylish-haskell` (that was enabled in my configuration) put constructors before the function.
I could disable it and undo changes in imports.

swamp-agr replied:
Fixed.

maksbotan commented on servant-client/src/Servant/Client/Internal/HttpClient.hs:
```diff
@@ -65,7 +65,7 @@ import           GHC.Generics
                  (renderHeader)
 import           Network.HTTP.Types
-                 (hContentType, renderQuery, statusCode, Status)
+                 (Status, hContentType, renderQuery, statusCode)
```
Same here.

swamp-agr replied:
`stylish-haskell` (that was enabled in my configuration) put constructors before the function.
I could disable it and undo changes in imports.

maksbotan replied:
Well, I think it makes sense to undo all changes to files where there are no significant changes now. Formatting files you touch for other reasons is fine.

swamp-agr replied:
Fixed.

maksbotan commented on servant-docs/src/Servant/Docs/Internal.hs:
```diff
@@ -161,6 +161,20 @@ data DocQueryParam = DocQueryParam
+  , _fragDesc   :: String -- user supplied
+  } deriving (Eq, Ord, Show)
+
+-- | There should be at most one `Fragment` per API endpoint.
```
```suggestion
-- | There should be at most one 'Fragment' per API endpoint.
```

swamp-agr replied:
Fixed.

maksbotan commented on servant-docs/src/Servant/Docs/Internal.hs:
```diff
@@ -161,6 +161,20 @@ data DocQueryParam = DocQueryParam
+-- | There should be at most one `Fragment` per API endpoint.
+-- So here we are keeping the first occurrence.
+combineFragment :: Maybe DocFragment -> Maybe DocFragment -> Maybe DocFragment
+Nothing `combineFragment` mdocFragment = mdocFragment
```
This is equal (i think) to `combineFragment a b = getFirst $ (First a) <> (First b)`

It's up to you to decide if you want to make this change.

swamp-agr replied:
I left this change as it was before. 
If you insist, I could definitely change it.

maksbotan commented on servant/src/Servant/API.hs:
```diff
@@ -19,6 +19,8 @@ module Servant.API (
   module Servant.API.QueryParam,
   -- | Retrieving parameters from the query string of the 'URI': @'QueryParam'@
+  module Servant.API.Fragment,
+  -- | Retrieving parameter from the fragment of the 'URI': @'Fragment'@
```
Hmm, now you can't /retrieve/ anything, maybe instead write here "Documenting the fragment part of the URI: ..." ?

swamp-agr replied:
Fixed.

maksbotan commented on servant/src/Servant/API/Fragment.hs:
```diff
@@ -0,0 +1,31 @@
+                 (Typeable)
+import           Servant.API.Modifiers
+
+-- | Lookup the value associated to the URI fragment
```
Same here: you don't lookup anything now as far as I understand, so the doc is misleading.

Furthermore, the modifiers are not needed anymore, right?

swamp-agr replied:
Fixed.

maksbotan commented on servant/src/Servant/API/TypeLevel.hs:
```diff
@@ -241,18 +249,53 @@ We might try to factor these our more cleanly, but the type synonyms and type
+instance OnlyOneFragment (Verb m s ct typ)
+
+type family FragmentUnique api :: Constraint where
+  FragmentUnique (sa :<|> sb)       = Or (FragmentUnique sa) (FragmentUnique sb)
```
Why `Or` and not `And`? 🤔

swamp-agr replied:
- `Or` here represents difference between several endpoints. 
- The fragment uniqueness check should be applied only per particular API endpoint. Not to the whole API.

maksbotan replied:
```haskell
type FailApi =
  (Fragment "foo" Text :> Fragment "bar" Int :> Get '[JSON] Int)
  :<|> (Fragment "baz" String) :> Get '[JSON] String)
```

With `Or` your check will pass for this API because it passes for the second branch, ignoring the error in the first one.

Sorry if I'm getting something wrong :)

swamp-agr replied:
Fixed.

maksbotan commented on servant-server/src/Servant/Server/Internal.hs:
```diff
@@ -880,5 +882,31 @@ type HasServerArrowTypeError a b =
+         , SBoolI (FoldLenient mods)
+         )
+#else
+instance (HasServer api context, FromHttpApiData a1, SBoolI (FoldLenient mods))
```
`FoldLenient` is not needed anymore, right?

swamp-agr replied:
Fixed.

maksbotan commented on servant/src/Servant/API/Modifiers.hs:
```diff
@@ -131,7 +132,8 @@ type RequestArgument mods a =
        (Maybe (If (FoldLenient mods) (Either Text a) a))
 
-
+type LenientArgument mods a =
```
This is not actually used? If so, I think you should just remove it.

swamp-agr replied:
Fixed.

maksbotan reviewed (commented):
Looks great, thanks! Left a couple of minor comments.

Otherwise I'm in favor of merging this. One thing that bothers me is the change in the `Link` datatype: can we go with a minor release or should it wait for 0.19? cc @fisx @alpmestan.

swamp-agr commented:
## Summary from review

- servant-docs: change backticks to apostrophes.
- servant:
    - [X] rephrase haddock section in import.
    - [X] remove `Fragment'`.
    - [X] remove `FoldLenient` usage.
    - [X] remove `LenientArgument`.
    - [X] adopt changes after fixes above.

- several things left (explanations provided):
    - imports.
    - `Or` vs. `And`.
    - `combineFragment`.

maksbotan commented on servant/src/Servant/API/TypeLevel.hs:
```diff
@@ -241,18 +249,53 @@ We might try to factor these our more cleanly, but the type synonyms and type
+-- ...Only one Fragment allowed per endpoint in api...
+-- ...
+-- ...In the instance declaration for...
+instance OnlyOneFragment (Verb m s ct typ)
```
btw, I just realized you need this instance for `UVerb` as well, right?

swamp-agr replied:
Fixed.

maksbotan commented on servant-server/src/Servant/Server/Internal.hs:
```diff
@@ -880,5 +882,29 @@ type HasServerArrowTypeError a b =
+    => HasServer (Fragment a1 :> api) context where
+  type ServerT (Fragment a1 :> api) m = ServerT api m
+
+  route Proxy context subserver = route (Proxy :: Proxy api) context subserver 
```
You can eta-reduce to have `route _ = route (Proxy :: Proxy api)`, similarly for `hoist`.

swamp-agr replied:
Fixed.

maksbotan commented on servant-server/src/Servant/Server/Internal.hs:
```diff
@@ -880,5 +882,29 @@ type HasServerArrowTypeError a b =
+#ifdef HAS_TYPE_ERROR
+instance (OnlyOneFragment api, HasServer api context, FromHttpApiData a1)
+#else
+instance (HasServer api context, FromHttpApiData a1)
```
Hmm... is `FromHttpApiData` actually needed here? Same for `ToHttpApiData` in the client.

swamp-agr replied:
Fixed.

maksbotan commented on servant-server/src/Servant/Server/Internal.hs:
```diff
@@ -880,5 +882,29 @@ type HasServerArrowTypeError a b =
+instance (HasServer api context, FromHttpApiData a1)
+#endif
+    => HasServer (Fragment a1 :> api) context where
+  type ServerT (Fragment a1 :> api) m = ServerT api m
```
You require `OnlyOneFragment api`, shouldn't it be `OnlyOneFragment (Fragment a1 :> api)`? Otherwise you could generate `Fragment a :> (Fragment b :> ...)`, where the thing in `( )` passes the check. Or am I missing something?

swamp-agr replied:
Fixed.

swamp-agr commented:
## Summary from review iter. 2

- [X] `Or` to `And` cross-check performed.
- [X] `instance AtLeastOneFragment UVerb`
- [X] simplified HasServer/HasClient instances (eta-reduce)
- [X] removed `From/To-HttpApiData`
- [X] fixed `HasServer`/`HasClient` instances for `FailAPI` example.

maksbotan reviewed (approved):
OK, now I'm satisfied, thanks!

@fisx @alpmestan are you going to take a final look?

swamp-agr commented:
@fisx, @fizruk, @arianvp, @maksbotan: Let me share my experience of contributing to `servant` with you: https://an-pro.org/posts/11-servant-fragment.html 

It is a honour for me!
Take care of yourself!

Thanks and Regards,
@swamp-agr
